src/library/scala/collection/immutable/TreeSet.scala

```diff
@@ -228,12 +196,50 @@ final class TreeSet[A] private (private val tree: RB.Tree[A, Unit])(implicit val
 
   override def foreach[U](f: A => U) = RB.foreachKey(tree, f)
 
-  override def rangeImpl(from: Option[A], until: Option[A]): TreeSet[A] = newSet(RB.rangeImpl(tree, from, until))
-  override def range(from: A, until: A): TreeSet[A] = newSet(RB.range(tree, from, until))
-  override def from(from: A): TreeSet[A] = newSet(RB.from(tree, from))
-  override def to(to: A): TreeSet[A] = newSet(RB.to(tree, to))
-  override def until(until: A): TreeSet[A] = newSet(RB.until(tree, until))
+  override def rangeImpl(from: Option[A], until: Option[A]): TreeSet[A] = newSetOrSelf(RB.rangeImpl(tree, from, until))
+  override def range(from: A, until: A): TreeSet[A] = newSetOrSelf(RB.range(tree, from, until))
+  override def from(from: A): TreeSet[A] = newSetOrSelf(RB.from(tree, from))
+  override def to(to: A): TreeSet[A] = newSetOrSelf(RB.to(tree, to))
+  override def until(until: A): TreeSet[A] = newSetOrSelf(RB.until(tree, until))
 
   override def firstKey = head
   override def lastKey = last
+
+
+  private def sameCBF(bf: CanBuildFrom[_,_,_]): Boolean = {
+    bf match {
+      case tsb: TreeSet.TreeSetBuilder[_] if tsb.ordering == ordering => true
+      case _ => false
+    }
+  }
```

Comment:
You may want to take `WrappedCanBuildFrom` (created by `breakOut`) into account:

```
  private def sameCBF(bf: CanBuildFrom[_,_,_]): Boolean = {
    bf match {
      case tsb: TreeSet.TreeSetBuilder[_] if tsb.ordering == ordering => true
      case w: WrappedCanBuildFrom[_,_,_] =>
        sameCBF(w.unwrapped)
      case _ => false
    }
  }
```

Reply:
well spotted!

Reply:
added when I rebased to avoid conflicts. This was the only change other than the conflicts